### PR TITLE
Add files for mltrack ranking and fix bug in bert_ranker

### PR DESCRIPTION
**Pull request by elenaermakova**
1. Added files for mltrack ranking:

- reader `mltrack_reader.py` to read and to preprocess raw data

- config file `mltrack_ranker.json` for model training

- `ndcg.py` metric for scoring observed predictions

- notebook `mltrack_submit.ipynb` to get submission file

2. Bug in return_probas condition:
If `self.return_probas` is True, then error appears in the following line because `pred` dimensionality in this case is equal to 1. Moreover, `predictions` is a vector containing probabilities of the last class which doesn't make sense in case of non-binary classification.
https://github.com/deepmipt/DeepPavlov/blob/0214bd4ca2a676057d287a5011a20fb7ee2ddaeb/deeppavlov/models/bert/bert_ranker.py#L101

### CODE REVIEW

File: deeppavlov/dataset_readers/mltrack_reader.py [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+        data_path = expand_path(data_path)
+        train_fname = data_path / "train.tsv"
+        test_fname = data_path / "final.tsv"
+        train_data = self.build_train_data(train_fname)
+        valid_data = self.build_validation_data(train_fname, "valid")
+        test_data = self.build_validation_data(test_fname, "test")
+        dataset = {"train": train_data, "valid": valid_data, "test": test_data}
+        return dataset
+
+    def build_validation_data(self, name, data_type):
```
Comment (yoptar) [inline deeppavlov/dataset_readers/mltrack_reader.py]: Add typing please

File: deeppavlov/dataset_readers/mltrack_reader.py [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+        lines = []
+        with open(name, "r") as f:
+            reader = csv.reader(f, delimiter="\t", quotechar=None)
+            for line in reader:
+                lines.append(line)
```
Comment (yoptar) [inline deeppavlov/dataset_readers/mltrack_reader.py]: Maybe something like `lines = list(reader)`?

File: deeppavlov/dataset_readers/mltrack_reader.py [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+        test_fname = data_path / "final.tsv"
+        train_data = self.build_train_data(train_fname)
+        valid_data = self.build_validation_data(train_fname, "valid")
+        test_data = self.build_validation_data(test_fname, "test")
+        dataset = {"train": train_data, "valid": valid_data, "test": test_data}
+        return dataset
+
+    def build_validation_data(self, name, data_type):
+        lines = []
+        with open(name, "r") as f:
```
Comment (yoptar) [inline deeppavlov/dataset_readers/mltrack_reader.py]:
```suggestion
        with open(name, "r", encoding="utf8") as f:
```

File: deeppavlov/dataset_readers/mltrack_reader.py [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+from deeppavlov.core.commands.utils import expand_path
+from deeppavlov.core.common.registry import register
+from deeppavlov.core.data.dataset_reader import DatasetReader
+
+
+@register("mltrack_reader")
+class MLtrackReader(DatasetReader):
+    """The class to read the MLtrack dataset from files.
+
+    Please, see https://contest.yandex.ru/algorithm2018
```
Comment (yoptar) [inline deeppavlov/dataset_readers/mltrack_reader.py]: This url doesn't really clarify things

File: deeppavlov/dataset_readers/mltrack_reader.py [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+            for line in reader:
+                lines.append(line)
+        if data_type == "test":
+            for line in lines:
+                line.extend(["bad", 1])
+        pad = [""]
+        num = 0
+        y_voc = {"good": 2, "neutral": 1, "bad": 0}
+        samples, labels, label = [], [], []
+        cur_id = lines[0][0]
```
Comment (yoptar) [inline deeppavlov/dataset_readers/mltrack_reader.py]: Can the file be empty?

File: deeppavlov/dataset_readers/mltrack_reader.py [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+        y_voc = {"good": 2, "neutral": 1, "bad": 0}
+        for line in lines:
+            context = self.clean_data(" ".join(line[1:4]))
+            reply = self.clean_data(line[5])
+            samples.append([context, reply])
+            labels.append(y_voc[line[-2]])
+        return list(zip(samples, labels))
+
+    def clean_data(self, text):
+        """Clean or fix invalid symbols in given text"""
```
Comment (yoptar) [inline deeppavlov/dataset_readers/mltrack_reader.py]:
```suggestion
        """Clean or fix invalid symbols in a given text"""
```

File: deeppavlov/dataset_readers/mltrack_reader.py (hunk 1 above)
Comment (yoptar) [reply deeppavlov/dataset_readers/mltrack_reader.py]: everywhere

File: deeppavlov/configs/ranking/mltrack_ranker.json [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+    "val_every_n_epochs": 3,
+    "evaluation_targets": [],
+    "validate_first": false,
+    "tensorboard_log_dir": "{MODELS_PATH}/"
+  },
+  "metadata": {
+    "variables": {
+      "ROOT_PATH": "~/.deeppavlov",
+      "DOWNLOADS_PATH": "{ROOT_PATH}/downloads",
+      "MODELS_PATH": "{ROOT_PATH}/models/mltrack_model"
```
Comment (yoptar) [inline deeppavlov/configs/ranking/mltrack_ranker.json]: Please use default value for `"MODELS_PATH": "{ROOT_PATH}/models"` to conform to #844  
You can add another variable like `"MODEL_PATH": "{MODELS_PATH}/mltrack_model"`

File: deeppavlov/configs/ranking/mltrack_ranker.json [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+      },
+      {
+        "class_name": "bert_ranker",
+        "n_classes": 3,
+        "one_hot_labels": false,
+        "return_probas": false,
+        "bert_config_file": "{DOWNLOADS_PATH}/bert_model/rubert_cased_L-12_H-768_A-12_v1/bert_config.json",
+        "pretrained_bert": "{DOWNLOADS_PATH}/bert_model/rubert_cased_L-12_H-768_A-12_v1/bert_model.ckpt",
+        "save_path": "{MODELS_PATH}/model",
+        "load_path": "{MODELS_PATH}/model",
```
Comment (yoptar) [inline deeppavlov/configs/ranking/mltrack_ranker.json]: There is no model in downloads list. Do you not share a pretrained model?

File: deeppavlov/dataset_readers/mltrack_reader.py (hunk 4 above)
Comment (puleon) [reply deeppavlov/dataset_readers/mltrack_reader.py]: Please provide a more detailed description of the class

File: deeppavlov/dataset_readers/mltrack_reader.py (hunk 2 above)
Comment (elenaermakova) [reply deeppavlov/dataset_readers/mltrack_reader.py]: Now your version is used

File: deeppavlov/dataset_readers/mltrack_reader.py (hunk 1 above)
Comment (elenaermakova) [reply deeppavlov/dataset_readers/mltrack_reader.py]: Added

File: deeppavlov/dataset_readers/mltrack_reader.py (hunk 5 above)
Comment (elenaermakova) [reply deeppavlov/dataset_readers/mltrack_reader.py]: This reader was made to preprocess specific data for mltrack competition so files can't be empty

File: deeppavlov/configs/ranking/mltrack_ranker.json (hunk 7 above)
Comment (elenaermakova) [reply deeppavlov/configs/ranking/mltrack_ranker.json]: Fixed

File: deeppavlov/configs/ranking/mltrack_ranker.json (hunk 8 above)
Comment (elenaermakova) [reply deeppavlov/configs/ranking/mltrack_ranker.json]: If you mean a model that I was trained using this config, you can find it here in downloads list. I called it fine-tuned because this config allows to fine-tune weights from pretrained bert.  
https://github.com/deepmipt/DeepPavlov/blob/712717a268329015b41de1e60adb84531d1a490c/deeppavlov/configs/ranking/mltrack_ranker.json#L81-L82

File: deeppavlov/dataset_readers/mltrack_reader.py (hunk 4 above)
Comment (elenaermakova) [reply deeppavlov/dataset_readers/mltrack_reader.py]: I've changed url and extended the description.

File: deeppavlov/configs/ranking/mltrack_ranker.json (hunk 8 above)
Comment (yoptar) [reply deeppavlov/configs/ranking/mltrack_ranker.json]: But the path is different. So you can't just do `model = build_model('mltrack_ranker', download=True)` and get a pretrained model.

**Comment by yoptar** [comment]
Probably wont be merged any time soon. We can reopen this PR later if anything changes
